### ISSUE TITLE
TypeEvaluator crashes due to Protocol usage with TypeVar(infer_variance=True)

### ISSUE THREAD

**Issue by minmax** [microsoft/pyright#8137]
pyright 1.1.367
python 3.11

After upgrade from 1.1.366, pyright crashes with error:
```
An internal error occurred while type checking file "...": Error: Debug Failure. False expression: Expected variance to be inferred
```

Code to reproduce:
```python
from typing import Protocol
from typing_extensions import TypeVar

T = TypeVar("T", infer_variance=True)

class SupportsRead(Protocol[T]):
    def read(self, __length: int = ..., /) -> T: ...

def read(data: SupportsRead[bytes]) -> None: ...

read(open("/dev/null", "rb"))
```

**Comment by bryanforbes** [comment]
This also happens with PEP-695 syntax

**Comment by erictraut** [comment]
Thanks for the bug report and the repro, and apologies for the regression. I'm surprised that none of the pyright unit tests caught this. It turns out that it occurs only under very specific conditions. It was a long-standing bug, but it was unmasked by a recent (seemingly unrelated) code change.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.368.
